File: Middle of the Linked List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func middleNode(head *ListNode) *ListNode {
    slow := head
    fast := head.Next

    for fast != nil {
        fast = fast.Next
        slow = slow.Next

        if fast == nil {
            return slow
        }

        fast = fast.Next
    }
    return slow
}

File: Middle of the Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        auto slow = head;
        auto fast = head->next;

        while (fast) {
            fast = fast->next;
            slow = slow->next;

            if (!fast) return slow;

            fast = fast->next;
        }
        return slow;
    }
};
